src/renderer/viz/expressions/ramp.js

```diff
@@ -325,6 +342,35 @@ export default class Ramp extends BaseExpression {
         gl.uniform1f(this._getBinding(program).keyMinLoc, (this.minKey));
         gl.uniform1f(this._getBinding(program).keyWidthLoc, (this.maxKey) - (this.minKey));
         drawMetadata.freeTexUnit++;
+
+        if (this.input.type === 'category' && this.input.isA(Property)) {
+            gl.activeTexture(gl.TEXTURE0 + drawMetadata.freeTexUnit);
+            if (this._translatedIds !== this._metadata.properties[this.input.name].categories.length) {
+                this._translatedIds = this._metadata.properties[this.input.name].categories.length;
+                this._translateTexture = gl.createTexture();
+                const translatorPixels = new Float32Array(256 * 256);
+                for (let i = 0; i < this._metadata.properties[this.input.name].categories.length; i++) {
+                    const id = this._metadata.categoryToID.get(this._metadata.properties[this.input.name].categories[i].name);
+                    const value = i / (this._metadata.properties[this.input.name].categories.length - 1);
+                    const vec2Id = {
+                        x: id % 256,
+                        y: Math.floor(id / 256)
+                    };
+                    translatorPixels[256 * vec2Id.y + vec2Id.x] = value;
+                }
+                gl.bindTexture(gl.TEXTURE_2D, this._translateTexture);
```

Comment:
I think we can reuse `_bindGLTexture` method, which also checks the ramp type.

We can add a `texture` parameter to the method:

```js
_bindGLTexture (gl, texturePixels, texture) {
        gl.bindTexture(gl.TEXTURE_2D, texture);
        ...
}
```

And call it here as follows:

```js
this._bindGLTexture(gl.TEXTURE_2D, translatorPixels, this._translateTexture);
```

What do you think?

Reply:
We are currently using `this._bindGLTexture(gl, texturePixels)` in `_computeGLTextureIfNeeded` method. We can call it there like this:

* `this._bindGLTexture(gl, texturePixels, this.texture)`

And here:

* `this._bindGLTexture(gl, translatorPixels, this._translateTexture)`

Reply:
:confused: I don't understand, could you expand the implementation of `_bindGLTexture`?

Reply:
Yep, sorry!

```js
    _bindGLTexture (gl, texturePixels, texture) {
        gl.bindTexture(gl.TEXTURE_2D, texture);

        if (this.type === rampTypes.COLOR) {
            gl.pixelStorei(gl.UNPACK_PREMULTIPLY_ALPHA_WEBGL, false);
            gl.texImage2D(gl.TEXTURE_2D, 0, gl.RGBA, COLOR_ARRAY_LENGTH, 1, 0, gl.RGBA, gl.UNSIGNED_BYTE, texturePixels);
            gl.texParameteri(gl.TEXTURE_2D, gl.TEXTURE_MIN_FILTER, gl.LINEAR);
            gl.texParameteri(gl.TEXTURE_2D, gl.TEXTURE_MAG_FILTER, gl.LINEAR);
        } else {
            gl.texImage2D(gl.TEXTURE_2D, 0, gl.ALPHA, COLOR_ARRAY_LENGTH, 1, 0, gl.ALPHA, gl.FLOAT, texturePixels);
            gl.texParameteri(gl.TEXTURE_2D, gl.TEXTURE_MIN_FILTER, gl.NEAREST);
            gl.texParameteri(gl.TEXTURE_2D, gl.TEXTURE_MAG_FILTER, gl.NEAREST);
        }

        gl.texParameteri(gl.TEXTURE_2D, gl.TEXTURE_WRAP_S, gl.CLAMP_TO_EDGE);
        gl.texParameteri(gl.TEXTURE_2D, gl.TEXTURE_WRAP_T, gl.CLAMP_TO_EDGE);
    }
```

So, instead of 

```js
      gl.bindTexture(gl.TEXTURE_2D, this._translateTexture);
      gl.pixelStorei(gl.UNPACK_PREMULTIPLY_ALPHA_WEBGL, false);
      gl.texImage2D(gl.TEXTURE_2D, 0, gl.ALPHA, 256, 256, 0, gl.ALPHA, gl.FLOAT, translatorPixels);
      gl.texParameteri(gl.TEXTURE_2D, gl.TEXTURE_MIN_FILTER, gl.NEAREST);
      gl.texParameteri(gl.TEXTURE_2D, gl.TEXTURE_MAG_FILTER, gl.NEAREST);
      gl.texParameteri(gl.TEXTURE_2D, gl.TEXTURE_WRAP_S, gl.CLAMP_TO_EDGE);
      gl.texParameteri(gl.TEXTURE_2D, gl.TEXTURE_WRAP_T, gl.CLAMP_TO_EDGE);
``` 
We can call:

```js
this._bindGLTexture(gl, translatorPixels, this._translateTexture)
```